File: client/cpp/kserve-api/samples/grpc_infer_dummy.cc
//*****************************************************************************
// Copyright 2022 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************
#include <iostream>
#include <string>

#include <cxxopts.hpp>

#include "grpc_client.h"

namespace tc = triton::client;

#define FAIL_IF_ERR(X, MSG)                                              \
    {                                                                    \
        tc::Error err = (X);                                             \
        if (!err.IsOk()) {                                               \
            std::cerr << "error: " << (MSG) << ": " << err << std::endl; \
            exit(1);                                                     \
        }                                                                \
    }

namespace {

void ValidateShapeAndDatatype(
    const std::string& name, std::shared_ptr<tc::InferResult> result) {
    std::vector<int64_t> shape;
    FAIL_IF_ERR(
        result->Shape(name, &shape), "unable to get shape for '" + name + "'");
    // Validate shape
    if ((shape.size() != 2) || (shape[0] != 1) || (shape[1] != 10)) {
        std::cerr << "error: received incorrect shapes for '" << name << "'"
                  << std::endl;
        exit(1);
    }
    std::string datatype;
    FAIL_IF_ERR(
        result->Datatype(name, &datatype),
        "unable to get datatype for '" + name + "'");
    // Validate datatype
    if (datatype.compare("FP32") != 0) {
        std::cerr << "error: received incorrect datatype for '" << name
                  << "': " << datatype << std::endl;
        exit(1);
    }
}

}  // namespace

int main(int argc, char** argv) {
    cxxopts::Options opt("grpc_infer_dummy", "Sends requests via KServe gRPC API.");

    // clang-format off
    opt.add_options()
    ("h,help", "Show this help message and exit")
    ("grpc_address", "Specify url to grpc service. default:localhost", cxxopts::value<std::string>()->default_value("localhost"))
    ("grpc_port", "Specify port to grpc service. default:9000", cxxopts::value<std::string>()->default_value("9000"))
    ("input_name", "Specify input tensor name. default: input", cxxopts::value<std::string>()->default_value("input"))
    ("output_name", "Specify input tensor name. default: output", cxxopts::value<std::string>()->default_value("output"))
    ("model_name", "Define model name, must be same as is in service. default: dummy", cxxopts::value<std::string>()->default_value("dummy"))
    ("model_version", "Define model version.")
    ("timeout", "", cxxopts::value<int>()->default_value("0"))
    ;
    // clang-format on

    auto args = opt.parse(argc, argv);

    if (args.count("help")) {
        std::cout << opt.help() << std::endl;
        exit(0);
    }

    std::string url(args["grpc_address"].as<std::string>() + ":" + args["grpc_port"].as<std::string>());
    std::string model_name = args["model_name"].as<std::string>();

    // Create a InferenceServerGrpcClient instance to communicate with the
    // server using gRPC protocol.
    std::unique_ptr<tc::InferenceServerGrpcClient> client;

    FAIL_IF_ERR(
        tc::InferenceServerGrpcClient::Create(&client, url),
        err);

    std::vector<float> inputb_data(10);
    for (size_t i = 0; i < 10; ++i) {
        inputb_data[i] = i;
    }

    std::vector<int64_t> shape{1, 10};

    // Initialize the inputs with the data.
    tc::InferInput* inputb;

    FAIL_IF_ERR(
        tc::InferInput::Create(&inputb, "b", shape, "FP32"),
        "unable to get b");
    std::shared_ptr<tc::InferInput> inputb_ptr;
    inputb_ptr.reset(inputb);

    FAIL_IF_ERR(
        inputb_ptr->AppendRaw(
            reinterpret_cast<uint8_t*>(&inputb_data[0]),
            inputb_data.size() * sizeof(float)),
        "unable to set data for b");

    // Generate the outputs to be requested.
    tc::InferRequestedOutput* outputa;

    FAIL_IF_ERR(
        tc::InferRequestedOutput::Create(&outputa, "a"),
        "unable to get 'a'");
    std::shared_ptr<tc::InferRequestedOutput> outputa_ptr;
    outputa_ptr.reset(outputa);

    tc::InferOptions options(model_name);
    if (args.count("model_version"))
        options.model_version_ = args["model_version"].as<std::string>();
    options.client_timeout_ = args["timeout"].as<int>();

    std::vector<tc::InferInput*> inputs = {inputb_ptr.get()};
    std::vector<const tc::InferRequestedOutput*> outputs = {outputa_ptr.get()};

    tc::InferResult* results;
    FAIL_IF_ERR(
        client->Infer(&results, options, inputs, outputs),
        "unable to run model");
    std::shared_ptr<tc::InferResult> results_ptr;
    results_ptr.reset(results);

    // Validate the results...
    ValidateShapeAndDatatype("a", results_ptr);

    // Get pointers to the result returned...
    float* outputa_data;
    size_t outputa_byte_size;
    FAIL_IF_ERR(
        results_ptr->RawData(
            "a", (const uint8_t**)&outputa_data, &outputa_byte_size),
        "unable to get result data for 'a'");
    if (outputa_byte_size != 40) {
        std::cerr << "error: received incorrect byte size for 'a': "
                  << outputa_byte_size << std::endl;
        exit(1);
    }

    for (size_t i = 0; i < 10; ++i) {
        std::cout << inputb_data[i] << " => "
                  << *(outputa_data + i) << std::endl;

        if ((inputb_data[i] + 1) != *(outputa_data + i)) {
            std::cerr << "error: Incorrect sum" << std::endl;
        }
    }

    tc::InferStat infer_stat;
    client->ClientInferStat(&infer_stat);
    std::cout << "======Client Statistics======" << std::endl;
    std::cout << "completed_request_count "
              << infer_stat.completed_request_count << std::endl;
    std::cout << "cumulative_total_request_time_ns "
              << infer_stat.cumulative_total_request_time_ns << std::endl;
    std::cout << "cumulative_send_time_ns "
              << infer_stat.cumulative_send_time_ns << std::endl;
    std::cout << "cumulative_receive_time_ns "
              << infer_stat.cumulative_receive_time_ns << std::endl;


    return 0;
}

File: client/java/kserve-api/src/main/java/clients/grpc_infer_dummy.java
//*****************************************************************************
// Copyright 2022 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************

package clients;

import java.nio.ByteOrder;
import java.nio.FloatBuffer;
import java.util.Arrays;
import java.util.List;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

import inference.GRPCInferenceServiceGrpc;
import inference.GRPCInferenceServiceGrpc.GRPCInferenceServiceBlockingStub;
import inference.GrpcPredictV2.InferTensorContents;
import inference.GrpcPredictV2.ModelInferRequest;
import inference.GrpcPredictV2.ModelInferResponse;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

public class grpc_infer_dummy {

	public static void main(String[] args) {
		Options opt = new Options();
		opt.addOption("h", "help", false, "Show this help message and exit");
		opt.addOption(Option.builder("a").longOpt("grpc_address").hasArg().desc("Specify url to grpc service. ")
				.argName("GRPC_ADDRESS").build());
		opt.addOption(Option.builder("p").longOpt("grpc_port").hasArg().desc("Specify port to grpc service. ")
				.argName("GRPC_PORT").build());
		opt.addOption(Option.builder("i").longOpt("input_name").hasArg().desc("Specify input tensor name. ")
				.argName("INPUT_NAME").build());
		opt.addOption(Option.builder("o").longOpt("output_name").hasArg().desc("Specify output tensor name. ")
				.argName("OUTPUT_NAME").build());
		opt.addOption(Option.builder("n").longOpt("model_name").hasArg()
				.desc("Define model name, must be same as is in service").argName("MODEL_NAME").build());
		opt.addOption(Option.builder("v").longOpt("model_version").hasArg().desc("Define model version. ")
				.argName("MODEL_VERSION").build());
		CommandLineParser parser = new DefaultParser();
		CommandLine cmd = null;
		try {
			cmd = parser.parse(opt, args);
		} catch (ParseException exp) {
			System.err.println("Parsing failed.  Reason: " + exp.getMessage());
			System.exit(1);
		}

		if (cmd.hasOption("h")) {
			HelpFormatter formatter = new HelpFormatter();
			formatter.printHelp("grpc_infer_dummy [OPTION]...", opt);
			System.exit(0);
		}

		ManagedChannel channel = ManagedChannelBuilder
				.forAddress(cmd.getOptionValue("a", "localhost"), Integer.parseInt(cmd.getOptionValue("p", "9000")))
				.usePlaintext().build();
		GRPCInferenceServiceBlockingStub grpc_stub = GRPCInferenceServiceGrpc.newBlockingStub(channel);

		// Generate the request
		ModelInferRequest.Builder request = ModelInferRequest.newBuilder();
		request.setModelName(cmd.getOptionValue("n", "dummy"));
		request.setModelVersion(cmd.getOptionValue("v", ""));

		// Input data
		List<Float> lst = Arrays.asList(0f, 1f, 2f, 3f, 4f, 5f, 6f, 7f, 8f, 9f);
		InferTensorContents.Builder input_data = InferTensorContents.newBuilder();
		input_data.addAllFp32Contents(lst);

		// Populate the inputs in inference request
		ModelInferRequest.InferInputTensor.Builder input = ModelInferRequest.InferInputTensor
				.newBuilder();
		String defaultInputName = "b";
		input.setName(cmd.getOptionValue("i", defaultInputName));
		input.setDatatype("FP32");
		input.addShape(1);
		input.addShape(10);
		input.setContents(input_data);

		request.addInputs(0, input);

		ModelInferResponse response = grpc_stub.modelInfer(request.build());

		// Get the response outputs
		FloatBuffer fb = response.getRawOutputContentsList().get(0).asReadOnlyByteBuffer()
				.order(ByteOrder.LITTLE_ENDIAN).asFloatBuffer();
		float[] op = new float[fb.remaining()];
		fb.get(op);

		for (int i = 0; i < op.length; i++) {
			System.out.println(
					Float.toString(lst.get(i)) + " => " +
							Float.toString(op[i]));
		}

		channel.shutdownNow();

	}
}